src/core/dev/rfs.cpp

```diff
@@ -103,44 +107,37 @@ inline void rfs::prepare_filter_detach(int &filter_counter, bool decrease_counte
     filter_counter = filter_iter->second.counter;
     // if we do not need to destroy rfs_rule, still mark this rfs as detached
     m_b_tmp_is_attached = (filter_counter == 0) && m_b_tmp_is_attached;
-    if (filter_counter != 0 || filter_iter->second.rfs_rule_vec.empty()) {
+    if (filter_counter != 0) {
         return;
     }
 
     BULLSEYE_EXCLUDE_BLOCK_START
-    if (m_attach_flow_data_vector.size() != filter_iter->second.rfs_rule_vec.size()) {
-        // sanity check for having the same number of qps on all rfs objects
-        rfs_logerr("all rfs objects in the ring should have the same number of elements");
+    if (m_rfs_flow && m_rfs_flow != filter_iter->second.rfs_rule_holder) {
+        rfs_logerr("our assumption that there should be only one rule for filter group is wrong");
+    } else if (filter_iter->second.rfs_rule_holder) {
+        m_rfs_flow = filter_iter->second.rfs_rule_holder;
+        rfs_logdbg("prepare_filter_detach copying rfs_flow, Tag: %" PRIu32
+                   ", Flow: %s, Ptr: %p, Counter: %d",
+                   m_flow_tag_id, m_flow_tuple.to_str().c_str(), m_rfs_flow,
+                   filter_iter->second.counter);
     }
     BULLSEYE_EXCLUDE_BLOCK_END
-
-    for (size_t i = 0; i < m_attach_flow_data_vector.size(); i++) {
-        BULLSEYE_EXCLUDE_BLOCK_START
-        if (m_attach_flow_data_vector[i]->rfs_flow &&
-            m_attach_flow_data_vector[i]->rfs_flow != filter_iter->second.rfs_rule_vec[i]) {
-            rfs_logerr(
-                "our assumption that there should be only one rule for filter group is wrong");
-        } else if (filter_iter->second.rfs_rule_vec[i]) {
-            m_attach_flow_data_vector[i]->rfs_flow = filter_iter->second.rfs_rule_vec[i];
-            rfs_logdbg("prepare_filter_detach copying rfs_flow, Tag: %" PRIu32
-                       ", Flow: %s, Index: %zu, Ptr: %p, Counter: %d",
-                       m_flow_tag_id, m_flow_tuple.to_str().c_str(), i,
-                       m_attach_flow_data_vector[i]->rfs_flow, filter_iter->second.counter);
-        }
-        BULLSEYE_EXCLUDE_BLOCK_END
-    }
 }
 
 rfs::rfs(flow_tuple *flow_spec_5t, ring_slave *p_ring, rfs_rule_filter *rule_filter /*= NULL*/,
          uint32_t flow_tag_id /*=0*/)
     : m_flow_tuple(rule_filter ? rule_filter->m_flow_tuple : *flow_spec_5t)
     , m_p_ring(p_ring)
+    , m_p_ring_simple(dynamic_cast<ring_simple *>(p_ring))
```

Comment:
[question] Can we create rfs not for ring_simple? In not, we don't need m_p_ring pointer.

Reply:
As far as i know we do. For example in ring_tap::attach_flow.